cpp/include/cuspatial/detail/trajectory.hpp

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuspatial/error.hpp>
+#include <cuspatial/trajectory.hpp>
+
+namespace cuspatial {
+namespace experimental {
+namespace detail {
+
+/**
+ * @brief Derive trajectories from points, timestamps, and object ids.
+ *
+ * Groups the input object ids to determine unique trajectories. Returns a
+ * table with the trajectory ids, the number of objects in each trajectory,
+ * and the offset position of the first object for each trajectory in the
+ * input object ids column.
+ *
+ * @param[in] x coordinates (km) (sorted by id, timestamp)
+ * @param[in] y coordinates (km) (sorted by id, timestamp)
+ * @param[in] object_id column of object (e.g., vehicle) ids
+ * @param[in] timestamp column (sorted by id, timestamp)
+ * @param[in] mr The optional resource to use for all allocations
+ * @param[in] stream Optional CUDA stream on which to schedule allocations
+ *
+ * @return an `std::pair<table, column>`:
+ *  1. table of (object_id, timestamp, x, y) sorted by (object_id, timestamp)
+ *  2. int32 column of end positions for each trajectory's last object
```

Comment:
The offsets are the position of the last object in each trajectory because that's what the legacy code was doing (using thrust `inclusive_scan` instead of `exclusive_scan`):

```
point_ids:    [0, 0, 0, 1, 1, 2]
traj_offsets: [2, 4, 5] // (0 is the inferred starting offset)
```

Initially I'd changed it to zero-based offsets, but as you [noted earlier](https://github.com/rapidsai/cuspatial/pull/171#discussion_r405960154), using inclusive_scan allows us to avoid returning the additional `lengths` buffer.

But now I think changing this back to zero-based offsets is the right approach, since it would allow us to replace the custom trajectory kernels with `cub::DeviceSegmentedReduce`.